ChainFilters Until Timestamp

ddd-mtl opened the pull request:
### Summary

Added new ChainFilter variant UntilTimestamp (from issue https://github.com/holochain/holochain/issues/4754)

### TODO:
- [ ] Check with core devs if actions can have identical timestamps
- [ ] Check with core devs if other areas need to be updated
- [ ] Implement unit tests
- [ ] CHANGELOGs updated with appropriate info
- [ ] All code changes are reflected in docs, including module-level docs

ThetaSinner commented:
> actions can have identical timestamps

Yes, they can. At some point in the past, action timestamps were required to increase between every action. In practice, the clock can return the same millisecond timestamp for two authored actions. It doesn't happen often but we'd seen sys validation rejecting actions for this. So, the restriction is now that timestamps must not decrease as the action sequence number increases.

This checked here https://github.com/holochain/holochain/blob/develop/crates/holochain/src/core/sys_validate.rs#L270. I notice that the documentation wasn't updated for the sys validation workflow when that change was made https://github.com/holochain/holochain/blob/develop/crates/holochain/src/core/workflow/sys_validation_workflow.rs#L49, that needs updating.

> other areas need to be updated

I haven't looked at the code yet but I'll just raise what comes to mind. The chain filtering logic is used in two places. One is through the `query` host function and the other is the poorly-named `get_agent_activity` host function. I would say that it's fine if this is only tested and working with the `query` host function but then we should open a new issue to decide whether to support this change for the other filter options. I'm not 100% that these two host functions share code, but from memory I think they share at least some things.

> unit tests

Yes please! Though I'm happy to have more discussion if that would be useful and you'd prefer to do that before writing tests.

ddd-mtl commented:
From what I can see `query` doesn't use `ChainFilters`, instead it uses its own thing: `ChainQueryFilter`
I need it for `must_get_agent_activity()` so I'm fine with that. Not sure how much it would be useful to also add it to `query`. Maybe  wait for a pull?

ddd-mtl commented:
Not sure about best way to query the sql db to get actions according to timestamp.
should I modify `action_hash_to_seq.sql` or create a new sql file for it.

jost-s commented:
> Not sure about best way to query the sql db to get actions according to timestamp. should I modify `action_hash_to_seq.sql` or create a new sql file for it?

A question without a mention is likely going to slip through. Please mention someone or `@holochain/holochain-devs` to call our attention.

ThetaSinner commented on crates/holochain/src/core/workflow/app_validation_workflow/run_validation_callback_tests.rs:
```diff
@@ -229,7 +229,7 @@ async fn validation_callback_awaiting_deps_agent_activity() {
                 let chain_filter = ChainFilter {
                     chain_top: delete.as_hash().clone(),
-                    filters: ChainFilters::Until(filter_hashes),
+                    stop_conditions: StopConditions::UntilHash(filter_hashes),
```
Perhaps `limit` or `bound`? To me `stop` suggests some exit or error condition but it's really more like an intented limit on the amount of data returned.

ddd-mtl replied:
Changed it to `limit`.

ThetaSinner commented on crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity.rs:
```diff
@@ -83,32 +83,48 @@ pub fn filter_then_check(
     filter: ChainFilter,
 ) -> StateQueryResult<Sequences> {
-    let mut statement = txn.prepare(ACTION_HASH_TO_SEQ)?;
+    let mut hash_statement = txn.prepare(ACTION_HASH_TO_SEQ)?;
```
`hash` is a verb, this reads more like a function name. Perhaps `by_hash_stmt` or `by_hash_statement` or even `hash_to_seq_statement`?

ddd-mtl replied:
Changed it to `by_hash_statement`

ThetaSinner commented on crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity.rs:
```diff
@@ -83,32 +83,48 @@ pub fn filter_then_check(
+    let mut ts_statement = txn.prepare(ACTION_TS_TO_SEQ)?;
+    let get_seq_from_ts = |ts: Timestamp| {
+        if let Some(scratch) = scratch {
+            if let Some(action) = scratch.actions().find(|a| a.get_timestamp() == ts) {
```
Timestamp equality? Intuitively, I'd expect a timestamp bound to include all the actions within the provided bound. The selected one for an action sequence would be the max or min depending on the direction of the equality. I can't make sense of an equality here.

ddd-mtl replied:
I've removed this part as I don't believe it actually makes sense to look in the scratch for old data. I had copied that over from how UntilHash was doing it.

ThetaSinner commented on crates/holochain_integrity_types/CHANGELOG.md:
```diff
@@ -7,8 +7,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ## Unreleased
 
+- **BREAKING**: Renamed enum `ChainFilters` to `StopConditions` and renamed some of its variants.
+- Adds `UntilTimestamp` variant to `ChainFilters`
```
```suggestion
- Add `UntilTimestamp` variant to `ChainFilters`
```

But this doesn't sound right, if the type has been renamed?

ThetaSinner commented on crates/holochain_integrity_types/CHANGELOG.md:
```diff
@@ -7,8 +7,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
 
+- **BREAKING**: Renamed enum `ChainFilters` to `StopConditions` and renamed some of its variants.
```
```suggestion
- **BREAKING**: Rename enum `ChainFilters` to `StopConditions` and rename some of its variants.
```

ThetaSinner replied:
The variant rename should be in the changelog.

Some people do update the individual changelogs but we're moving towards just maintaining all changes in the Holochain one. When we switch to the new release process, that's how it will work. So it's helpful to work that way now so we can better maintain our change history.

ddd-mtl replied:
Added variant changes in changelog. Still using this changelog for now and next time I'll use the main one,  if its okay?

jost-s reviewed (commented):
As a general comment, I prefer for breaking name changes to be in a separate PR and not mixed with a feature or bug fix.

It helps me as a reviewer if you point me to which tests you added.

Regarding the timestamp condition, I'm wondering if the precise timestamp of an action is always known when filtering, or if would be handier to have a `>=` in the `WHERE` clause instead?

jost-s commented:
> Not sure about best way to query the sql db to get actions according to timestamp.
> should I modify `action_hash_to_seq.sql` or create a new sql file for it?

I prefer a separate query like you created.

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,87 +16,96 @@ mod test;
+/// a given timestamp has been passed.
+/// Multiple `[`ActionHash`] can be provided. The filter will stop at and take the first one found.
+/// When providing a Timestamp, the filter will stop at and take the last item that has a
+/// Timestamp younger than the provided one. In the edge case of multiple actions having the same
```
```suggestion
/// Timestamp older than the provided one. In the edge case of multiple actions having the same
```

Or maybe it's "newer", I've not got to the database query logic yet.

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,87 +16,96 @@ mod test;
+/// Multiple `[`ActionHash`] can be provided. The filter will stop at and take the first one found.
+/// When providing a Timestamp, the filter will stop at and take the last item that has a
+/// Timestamp younger than the provided one. In the edge case of multiple actions having the same
+/// Timestamp as the stop condition, only one of the actions will be included.
```
This has to be more specific. It sounds like it could be either but it either needs to be well-defined as the one with the highest sequence number or we need to re-examine this logic and we should be including actions past the provided bound.

I'm inclined to think it's the latter. If you provide a timestamp bound, surely all timestamps within that bound should be included and there is no edge case here?

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,87 +16,96 @@ mod test;
     /// Allow all up to genesis.
     ToGenesis,
-    /// Take this many (inclusive of the starting position).
+    /// Take this many items (inclusive of the starting position).
```
```suggestion
    /// Take this many actions (inclusive of the starting position).
```

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,87 +16,96 @@ mod test;
-    /// Take this many (inclusive of the starting position).
+    /// Take this many items (inclusive of the starting position).
     Take(u32),
+    /// Continue until some timestamp is passed.
```
The chain is walked backwards, so this should be something like "All actions since the given timestamp"?

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,87 +16,96 @@ mod test;
 
-/// Create a deterministic hash to compare filters.
-impl<H: Eq + Ord + std::hash::Hash> core::hash::Hash for ChainFilters<H> {
+/// Create a deterministic hash to compare StopConditions.
```
```suggestion
/// Create a deterministic hash to compare [`StopConditions`].
```

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,87 +16,96 @@ mod test;
 
-/// Implement a deterministic partial eq to compare ChainFilters.
-impl<H: Eq + Ord + std::hash::Hash> core::cmp::PartialEq for ChainFilters<H> {
+/// Implement a deterministic partial eq to compare StopConditions.
```
```suggestion
/// Implement a deterministic partial eq to compare [`StopConditions`].
```

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,87 +16,96 @@ mod test;
 
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
-/// Input to the `must_get_agent_activity` call.
+/// Input of the `must_get_agent_activity` call.
```
"Input to" was valid

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -131,47 +141,89 @@ impl<H: Eq + Ord + std::hash::Hash> ChainFilter<H> {
+            StopConditions::ToGenesis => StopConditions::UntilTimestamp(timestamp),
+            StopConditions::Take(n) => StopConditions::Multiple(n, HashSet::new(), timestamp),
+            StopConditions::UntilTimestamp(old_ts) => {
+                StopConditions::UntilTimestamp(timestamp.max(old_ts))
```
This logic deserves a code comment I think

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -131,47 +141,89 @@ impl<H: Eq + Ord + std::hash::Hash> ChainFilter<H> {
         self
     }
 
+    /// Take all actions until this timestamp is passed.
```
Based on this doc comment, I'm still not clear what the boundary condition is. This is all an app developer would get when setting up a filter so this needs to be perfectly expliit about what condition is applied after using this function to build your chain filter.

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -131,47 +141,89 @@ impl<H: Eq + Ord + std::hash::Hash> ChainFilter<H> {
+        match &self.stop_conditions {
+            StopConditions::Take(n) => Some(*n),
+            StopConditions::Multiple(n, _, _) => {
+                if *n != u32::MAX {
```
This is a logic change to an existing piece of code that appears unrelated to the main change on this PR. This needs calling out so it doesn't get missed. I also don't understand why this change is being made, so the call out for it needs to include a justification.

ThetaSinner replied:
Looking much better as logic with `Option`, I'm finding that much easier to follow

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -131,47 +141,89 @@ impl<H: Eq + Ord + std::hash::Hash> ChainFilter<H> {
+        match &self.stop_conditions {
+            StopConditions::UntilTimestamp(ts) => Some(*ts),
+            StopConditions::Multiple(_, _, ts) => {
+                if ts.0 > 0 {
```
Oh I think I get it. This is handling default values. 

This seems really counterintuitive to me code-wise. I can see that it will produce the right value in this case. Though I'm not so sure about the limit logic above.

Why not just make the filter parameters optional?

ddd-mtl replied:
Changed the tuple to use Options. Code should be more intuitive now.

ThetaSinner commented on crates/holochain_sqlite/src/sql/cell/agent_activity/action_ts_to_seq.sql:
```diff
@@ -0,0 +1,9 @@
+  Action
+  JOIN DhtOp ON DhtOp.action_hash = Action.hash
+WHERE
+  DhtOp.authored_timestamp = :authored_timestamp
```
This again isn't doing at all what I'm understanding from the provided documentation so far. The logic I'm getting is that we walk the chain backwards until we find a boundary timestamp and then return everything later than it. This actually requires the user to know the timestamp of some action. How do they source that information? It could only come from the action and if you have the action then you know its hash and can use the existing hash bounding functionality.

I'd expect this to look like an `ORDER BY` and `WHERE DhtOp.authored_timestamp > :authored_timestamp`

ThetaSinner reviewed (changes requested):
Stopping my review here, I haven't looked at all the code but I'm finding the current implementation unclear and I think it needs discussion before we proceed.

If you have to know the timestamp, then you likely already know the hash. We need to have a shared understanding of what functionality this is trying to add before I can review it appropriately. Because this affecting the HDI and needs to be assessed for determinism, it's really important that this gets well reviewed.

I'm also wondering whether this is actually how you want to implement rate limiting. If that's the only purpose of this change, then rate-limiting is already partially built into Holochain and blocking people permantently is problematic. There are other strategies for rate limiting that don't involve permenant bans for what might well be an accident. I was under the impression that this change was affecting `get_agent_actiivty` and not `must_get_agent_activity`. I'd be much less concerned about that change and blocking agents temporarily in coordinator logic. Given the importance of this code's correctness to Holochain's design, I think we need to spend some time discussing and justifying the addition of this feature. Please could this be turned into a MM conversation?

ddd-mtl commented:
@ThetaSinner 
I renamed and reworked the query for getting the seq of the oldest action with the given timestamp (`ts_to_seq.sql`).
I refactored its use in `Sequences::find_sequences()` accordingly.
I refactored the `LimitConditions::Multiple` variant to use Options in its tuple instead of default values.
I changed the test code (`TestChainItem`) to make it possible for testing against timestamps that are not present in the chain, and also for testing the edge case of multiple items with the same timestamp.
Added more unit tests.
Did all the requested renamings and doc fix.

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,74 +16,83 @@ mod test;
+/// a given timestamp has been passed.
+/// Multiple [`ActionHash`] can be provided. The filter will stop at and take the first one found.
+/// When providing a Timestamp, the filter will stop at and take the oldest action that has a
+/// Timestamp newer than the provided one. In the edge case of multiple actions having the same
+/// Timestamp as the limit condition, the filter will stop at the action with the lowest sequence.
```
Just double checked myself on this, but I had mentioned elsewhere in the previous review that I thought we could eliminate the edge case logic so I am being consistent with myself here :)

ddd-mtl replied:
Sorry, I'm not sure I understand what you mean here. Indeed there is no edge case specific code. This comment is for informing the user how the code will behave in this situation. I think its useful info, but if you think its not as its implicit then I can remove this last comment no problem.

ThetaSinner replied:
I don't think it needs removing necessarily, just the reference to the edge case. 

I do think it's reasonably clear that actions at the same timestamp would be included but maybe it's a good thing to document anyway. Just as "In the case of" rather than "in the edge case of".

ddd-mtl replied:
changed as requested

ThetaSinner commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -131,47 +146,82 @@ impl<H: Eq + Ord + std::hash::Hash> ChainFilter<H> {
         self
     }
 
+    /// Take all actions since the given timestamp.
+    /// If a timestamp was already set, the biggest of the two is kept.
```
```suggestion
    /// Take all actions since the given timestamp.
    ///
    /// If a timestamp was already set, the biggest of the two is kept.
```

Minor thing but this affects the rendering in docs.rs. When you preview an item you see its first doc line and when you view the documentation for an item you see all of the content.

We have a huge number of doc comments that don't follow this style but it is the one recommended by the Rust documentation -> https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components

ThetaSinner replied:
```suggestion
    /// Take all actions since the given timestamp.
    /// If a timestamp was already set, the later of the two is kept.
```

ThetaSinner commented on crates/holochain_sqlite/src/sql/cell/agent_activity/ts_to_seq.sql:
```diff
@@ -0,0 +1,11 @@
+  DhtOp.authored_timestamp >= :until
+  AND DhtOp.type = :activity
+  AND Action.author = :author
+ORDER BY
```
I thought I'd commented before about this but I hadn't and I seem to have missed it in MM too. I'm seeing the logic now for why this doesn't need an upper bound on the query but it is returning a potentially long list of sequences where, looking at the hash version, it only really needs to return 1.

As you've already filtered for exactly the right condition with `>=` and sorted ascending, the value you want is actually the first on in the result list right? So this could be `LIMIT 1` and you have to return less from the database?

ddd-mtl replied:
changed as requested

ThetaSinner reviewed (commented):
Liking the changes, the `Option` is much clearer and the timestamp filter logic is looking sound now.

I have one suggestion that I think can improve the SQL query. I'm also questioning the scratch space change, I think that shouldn't have been removed and I've explained why.

Otherwise, I only have some really minor comments left.

ThetaSinner commented on crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity.rs:
```diff
@@ -83,32 +83,57 @@ pub fn filter_then_check(
+        // Search in scratch
+        let scratch_result = scratch.and_then(|s| {
+            s.actions()
+                .filter_map(|a| {
```
It's unclear whether we can rely on the ordering of actions in the scratch to be by sequence number. I think that is the case but it's not documented to be a requirement. 

With the actions in order, this is actually a `find_map` without the `min` and not as much needs to be iterated but this code is resilient to problem behavior in the scratch space so perhaps this is best!

ddd-mtl commented on crates/holochain_integrity_types/src/chain.rs:
```diff
@@ -16,74 +16,83 @@ mod test;
-/// this filter.
-pub enum ChainFilters<H: Eq + Ord + std::hash::Hash = ActionHash> {
+/// Specify when to stop walking down the chain.
+pub enum LimitConditions<H: Eq + Ord + std::hash::Hash = ActionHash> {
```
I'd be fine with `Limit` or `Bound` even though it could suggest a seq value, but if you don't mind at this stage I'd like to stick to the current proposal unless there is a good reason not to.

jost-s replied:
Fine by me, yes
